design/proposals/variable_schema_validation.md

```diff
@@ -0,0 +1,293 @@
+
+```text
+Message: Invalid type. Expected Integer but got String.
+Schema path: https://cyberark.com/mysql.schema.json#/properties/port/type
+```
+
+#### Secrets Encryption
+
+The JSON schema can contain a property called `non-secrets` which expects an array of JSON schema property names. The properties specified in this array will not be encrypted in the database, along with the JSON structure. The unspecified attributes, will by default get encrypted. This allows the protection of sensitive data while also allowing the non-sensitive values to be searchable. Meaning that the variable will not be encrypted as a whole, but only the sensitive attributes in the JSON structure.
+
+For the example above, the `mysql-db-creds` variable will be saved in the database as follows:
+
+```json
+{
+  "username": "my-user",
+  "password": "U2FsdGVkX19Ji6JpgnVHW3V47OtMJwuKi1Yf9nc0aP5QcuzdnIrpzZ2zMC90f24g",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+Before the variable value is returned to the client, the secret attributes are decrypted so that the given example above, would look like this:
+
+```json
+{
+  "username": "my-user",
+  "password": "my-password",
+  "address": "my-db.com",
+  "port": 3306
+}
+```
+
+#### Built-in Schemas
+
+To simplify the user experience, Conjur can come with prdefined JSON schemas for the most common secrets use cases. The list of predefined schmeas should include:
+
+- Database secrets - Oracle, MSSQL, MySQL, Postgres, MariaDB, DB2.
+- Cloud access keys - AWS, Azure, GCP.
+- X509 certificates.
+- JWTs
+
+Write about the advantages of atomicity, minimal changes in the database to implement this and so on.
+
+#### Automatic Permission Granting to Built-in Schemas
+
+In order to provide access for any role to the built-in schemas, a new built-in group should be introduced: `conjur/all`. All roles in the Conjur account should be added to this group automatically, thus keeping this group always up to date with all the roles that exist in the account.
+
+Usage example:
+
+```yaml
+- !permit
+  role: !group /conjur/all
+  privileges: [ read,execute ]
+  resource: !variable my-var
+```
+
+## Main Advantages and Expected Value From This Proposal
+
+- We leverage a standard way to enforce the content of the variable, with a well known JSON schema.
+- Tightly coupled values, such as username and password, are updated together in a single transaction. This will prevent momentary inconsistency in which each value was updated independently, one after the other.
+- The user will not have to look in the documentation or to understand the required variables, their structure or allowed input. Conjur will provide this feedback when an update attempt is made.
+- The feature does not require changes in the database schma, only in new built-in content.
+- The returned variable is in a JSON structure, same as the rest of our APIs responses.
+
+## Affected Areas
```

Comment:
What will be the complexity of 3rd parties to rotate those new complex variables password values ?

Reply:
I currently defined it as out of scope. If it will be important, we'll add support to that structure as well.

Reply:
please, elaborate about the complexity of 3rd parties rotation - maybe it worth to discuss it